### PR TITLE
7.x islandora 2012

### PR DESCRIPTION
**Pull request by qadan**
**JIRA Ticket**: https://jira.duraspace.org/browse/ISLANDORA-2012

# What does this Pull Request do?
Makes it so that the "generate_ocr" flag is used for OCR generation, and a new "generate_hocr" flag is used for HOCR generation.

# What's new?
* OCR/HOCR generation forced behaviour has been split between a 'generate_ocr' flag and a 'generate_hocr' flag
* An update hook has been added to apply current 'generate_ocr' flags on items to the 'generate_hocr' flag; that way, current behaviour is not interrupted on existing objects.

# How should this be tested?
The update hook should be tested to make sure that existing items have the 'generate_ocr' flag to them.
Other than that, most things to be tested UI/Drush-wise are elsewhere.

# Additional Notes:
* Despite being a fairly small update, this does fundamentally change the way we force OCR/HOCR-generating behaviour. Code that sets the 'generate_ocr' flag expecting to have that setting applied to HOCR now needs to have it split out. PRs are being made to paged content, book, and book batch, which will be linked to this.

# Interested parties
@jordandukart and @ajstanley maintain this so pinging e'm

### CODE REVIEW

File: includes/derivatives.inc
Comment (DiegoPino) [inline includes/derivatives.inc]: Should say "skipping generating"?  also, why is language being used to keep track of OCR flag?

**Comment by DiegoPino** [comment]
hu! @qadan can you think of a way of doing this in a different way? without adding a new predicate and iterating over all objects in a repository? Like setting a general flag per batch instead per object? It seems like to allow something simple and useful, the load is a bit too high. the RELS-EXT version count is already on the upper side so this could potentially add a lot more there. Not sure if people would be happy with https://github.com/Islandora/islandora_ocr/pull/81/files#diff-090ea4578007d1ac8bb83ef11b25f4dbR85 this running over a few million objects

We could discuss this change in the committers call. We deal with a lot of book pages and not sure if we are willing to handle this update, but open to discuss it of course

File: includes/derivatives.inc
Comment (whikloj) [reply includes/derivatives.inc]: Because we have an option in the select box to not perform OCR or choose the language.
https://github.com/Islandora/islandora_ocr/blob/7.x/includes/utilities.inc#L125

Comment (DiegoPino) [reply includes/derivatives.inc]: @whikloj  but $language == 'no_ocr' what language is that! kinda weird.. maybe the logic matches, but not sure if that makes sense at all. Anyway, out of scope of this right?

Comment (qadan) [reply includes/derivatives.inc]: Yeah, currently the global option counts for both. We could consider splitting these out in the future but I agree, a bit out of scope. Honestly a lot of the 'Generate/Don't Generate OCR' boxes are a bit misleading as they tie in HOCR without really informing

**Comment by qadan** [comment]
@DiegoPino Yeah I agree with that; I kind of struggled with this for a bit and that's a definite concern. My alternate idea - a bit uglier but skips the update batch and creates far fewer extra RELS-EXT versions - is to log the date/time the update occurred in the update hook as a Drupal variable, and update the logic for generating HOCR to include items with the `generate_ocr` flag whose `createdDate` is older than the logged date/time.

File: includes/derivatives.inc
Comment (adam-vessey) [inline includes/derivatives.inc]: Two copies of much the same functions now... should we parameterize? Possibly introduce constants for the flags? Deprecate the present functions, and make 'em call our parameterized version?

File: includes/derivatives.inc
Comment (adam-vessey) [inline includes/derivatives.inc:354]: Could dedup this to a single parameterized function... Could then avoid issues such as down on line 353 (referring to "OCR").

File: includes/derivatives.inc
Comment (adam-vessey) [inline includes/derivatives.inc]: Was this supposed to be replaced with the "remove hocr flag" dealio? (and the "remove ocr flag" dealio moved to the "derive ocr" bit?)

**Review by adam-vessey — COMMENTED** [review]
Copypasta seems to be spreading...

File: includes/derivatives.inc
Comment (adam-vessey) [inline includes/derivatives.inc]: Also, just noticed... It was doing the current thing previously, but this should probably use `NULL` to match all values, since the empty string _is_ a valid RDF plain literal value... Of course, to make use of 'em, [Tuque itself would have to be made to differentiate between `FALSE`-y values](https://github.com/Islandora/tuque/blob/1.x/FedoraRelationships.php#L397).

File: includes/derivatives.inc
Comment (qadan) [reply includes/derivatives.inc]: HAHA DISREGARD

... will fix

File: includes/derivatives.inc
Comment (qadan) [reply includes/derivatives.inc]: Well that's spooky

File: includes/derivatives.inc
Comment (adam-vessey) [reply includes/derivatives.inc]: This... is still relevant? Two functions which differ only by the relationship in question...

File: includes/derivatives.inc
Comment (adam-vessey) [inline includes/derivatives.inc]: In the case of the objects created before the update date wouldn't be hit here... The relationship would be removed when processing the OCR, so it wouldn't (/shouldn't) be [available to start the HOCR stuff](https://github.com/Islandora/islandora_ocr/pull/81/commits/a6b45ed4f56a94384079c0fc7c2b5508d4051115#diff-f20cd9f6981182cb8c80727774d28d4fR350).

File: islandora_ocr.install
Comment (adam-vessey) [inline islandora_ocr.install]: I'm... almost inclined to suggested making sure there is anything present that this will effect, since the relationship is removed after generation... It only persists when generation is being prevented.

File: includes/derivatives.inc
Comment (adam-vessey) [inline includes/derivatives.inc]: Would probably be safer just to make it copy the value over, instead of trying to cross wires... Though... not sure where would be the best place... Would have to be where the OCR one is sure to still exist...

File: includes/derivatives.inc
Comment (adam-vessey) [inline includes/derivatives.inc]: `$type` and `$rels_flag` are co-dependent, yeah? Probably better to eliminate one and look up the other... Possibly build up an array with the relevant info to use later?... Something like:

```php
function islandora_ocr_derive_datastream($type, AbstractObject $object, $force = FALSE) {
  $map = array(
    'HOCR' => array(
      'function' => 'islandora_ocr_create_hocr',
      'flag' => ISLANDORA_OCR_GENERATE_HOCR_RELS_STATEMENT,
    ),
    'OCR' => array(
      'functions' => 'islandora_ocr_create_ocr',
      'flag' => ISLANDORA_OCR_GENERATE_OCR_RELS_STATEMENT,
    ),
  );

  if ($updated_date = variable_get(ISLANDORA_OCR_UPDATE_7001_TIMESTAMP, FALSE)) {
    // Copy ISLANDORA_OCR_GENERATE_OCR_RELS_STATEMENT to ISLANDORA_OCR_GENERATE_HOCR_RELS_STATEMENT for objects that still have the flag and were created when only the former existed.
    // ... Could possibly isolate this to a separate function, just passing in the object?
    $diff = date_diff(new DateTime($timestamp), $object->createdDate);
    // Negative diffs indicate an object created before the timestamp.
    if ($diff->invert) {
      $rels = $object->relationships;
      $initial_autocommit = $rels->autoCommit;
      foreach ($rels->get(ISLANDORA_RELS_EXT_URI, ISLANDORA_OCR_GENERATE_OCR_RELS_STATEMENT)) as $result) {
        $rels->autoCommit = FALSE;
        $rels->add(ISLANDORA_RELS_EXT_URI, ISLANDORA_OCR_GENERATE_OCR_RELS_STATEMENT, $result['object']['value'], RELS_TYPE_PLAIN_LITERAL);
      }
      $rels->autoCommit = $initial_autocommit;
    }
  }

  if (!isset($map[$type)) {
    throw new InvalidArgumentException(t('Invalid OCR datastream type "@type"', array(
      '@type' => $type,
    )));
  }
  $rels_flag = $map[$type]['flag'];
  $function = $map[$type]['function'];

  [...]
  
  $derivative_file = $function($file, array('language' => $language));

  [...]

  _islandora_ocr_remove_generation_rels_ext_statement($object, $rels_flag);
}

[...]

function _islandora_ocr_remove_generation_rels_ext_statement(AbstractObject $object, $flag) {
  $rels = $object->relationships;
  if ($rels->get(ISLANDORA_RELS_EXT_URI, $flag, NULL, RELS_TYPE_PLAIN_LITERAL)) {
    $rels->remove(ISLANDORA_RELS_EXT_URI, $flag, NULL, RELS_TYPE_PLAIN_LITERAL);
  }
}
```

File: includes/derivatives.inc
Comment (adam-vessey) [inline includes/derivatives.inc]: [Shaaaaaaaaaame](https://travis-ci.org/Islandora/islandora_ocr/jobs/255025004#L332). :V

**Review by adam-vessey — CHANGES_REQUESTED** [review]
Seems to be some issues with the logic... also, a syntax error.

File: includes/derivatives.inc
Comment (qadan) [reply includes/derivatives.inc]: boof, whoops

File: includes/derivatives.inc
Comment (adam-vessey) [inline includes/derivatives.inc]: This could just call `_islandora_ocr_remove_generating_rels_ext_statement` now, yeah?

File: islandora_ocr.install
Comment (adam-vessey) [inline islandora_ocr.install]: Not terribly concerned, but this could be slightly more efficiently written... be run as either an ["ask" query](https://www.w3.org/TR/rdf-sparql-query/#ask) (which returns a single result with a single variable ("k0" or "k1" or something of the like) containing either the string "true" or "false"), or using the `ri->countQuery()` bit... Could even add the "LIMIT" directly to the query, so it can avoid building up the entire projection before returning only the first... or possibly combine the last two, with the `LIMIT` and the `countQuery()` business.

File: islandora_ocr.install
Comment (adam-vessey) [inline islandora_ocr.install]: Could... just avoid setting? The [default is `FALSE`](https://github.com/Islandora/islandora_ocr/pull/81/commits/1534944cb14dce48372719ffb6ab816ae7106e62#diff-f20cd9f6981182cb8c80727774d28d4fR279), yeah? (or [`FALSE`-y](https://github.com/Islandora/islandora_ocr/pull/81/commits/1534944cb14dce48372719ffb6ab816ae7106e62#diff-f20cd9f6981182cb8c80727774d28d4fR365))

File: includes/derivatives.inc
Comment (adam-vessey) [reply includes/derivatives.inc:354]: Is did.

**Review by adam-vessey — DISMISSED** [review]
Really, just the change to use the new helper in the deprecated function... Could live with the other stuff.

File: islandora_ocr.install
Comment (qadan) [reply islandora_ocr.install]: yee

File: includes/derivatives.inc
Comment (qadan) [reply includes/derivatives.inc]: actually this is unnecessary now that it's being done above. forgot to remove

File: includes/derivatives.inc
Comment (willtp87) [inline includes/derivatives.inc]: Assignment?

**Comment by jordandukart** [comment]
Bump @dannylamb @Islandora/7-x-1-x-committers this has been open without movement for a month.

**Comment by DiegoPino** [comment]
@jordandukart sorry but I'm really not for adding more predicates and more versions to rels-ext to keep derivative flags around. I really never liked the idea, not the purpose of rels-ext to deal with this. So i have really really to think about this. Not longer a coding issue (code looks ok now, still needs testing) but more good practices question

**Comment by adam-vessey** [comment]
@DiegoPino: It sounds more like you have an issue with the existing practices to which we are adhering--introducing a flag for the HOCR for additional specificity, parallel with the existing (and now more specific) for OCR--rather than the adherence itself. We've avoided running the repo-wide batch; however, a move away from the present use of RELS-EXT statements for object-level properties in this way _would_ necessitate a repository-wide batch in order to pull the statements out of RELS-EXT into whatever other datastream(/storage mechanism), and still end up creating additional versions of the RELS-EXT with their removal... In any case, developing an alternative to handle this state is outside of the scope of the ticket: We're adding specificity by adding a single additional predicate, not developing new object-level state storage.

With this PR, it looks like there should only be a single additional RELS-EXT version incurred for new objects which get OCR, when the flag is removed--the `autocommit=FALSE` behaviour should have them both be added to objects within a single version (if they do not in fact get added to the RELS-EXT prior to object ingest, before versioning is relevant).

Similarly, I'm not sure the assertion that (paraphrasing): "it's not the purpose of RELS-EXT to deal with arbitrary object-level properties" is correct... If this was true, then there would be no need to allow the use of arbitrary literals or URIs as the object in the RELS-EXT, nor allowing arbitrary predicates. They [_do_ indicate](https://wiki.duraspace.org/pages/viewpage.action?pageId=66585838#DigitalObjectRelationships-RELS-EXTValidation):

> [RELS-EXT] is intended to be dedicated to solely object-to-object relationships and not used to make general descriptive assertions about objects

"general descriptive assertions", I interpret more to indicate things like "descriptive metadata", with which I agree entirely: It should be packed up in its own datastreams like DC and MODS and so on... In our instance, we're specifically "describing" machine-readable state. As for object-to-object relationships... we _could_ create discrete objects to represent the states; however, this seems... excessive? ... And still requiring of repo-wide updates and additional versions if we wanted to follow through at this time.

File: includes/derivatives.inc
Comment (DiegoPino) [inline includes/derivatives.inc]: Why is this an internal "_" function but the (new) setter is a global available one? I

File: islandora_ocr.module
Comment (DiegoPino) [inline islandora_ocr.module]: Where is this used? ISLANDORA_OCR_UPDATE_7001_TIMESTAMP

**Review by DiegoPino — COMMENTED** [review]
I have a few questions

File: includes/derivatives.inc
Comment (adam-vessey) [reply includes/derivatives.inc]: This is consistent with how they're used. This function (`_islandora_ocr_remove_generating_rels_ext_statement()`) is only used (or intended for, in any case) inside of this module... The setter (`islandora_ocr_set_generating_rels_ext_statements()`) is called externally, in all the related pull requests:
* [`islandora_paged_content`](https://github.com/Islandora/islandora_paged_content/pull/131/files#diff-513c241a5c2f213d72b035bc1baa0743R805)
* [`islandora_solution_pack_book`](https://github.com/Islandora/islandora_solution_pack_book/pull/196/files#diff-25b6cf4bd16ae6912ebb116b013bc6a5R286)
* [`islandora_book_batch`](https://github.com/Islandora/islandora_book_batch/pull/68/files#diff-9d74701071b57981e86c31d3ea01aad0R480); and,
* [`islandora_newspaper_batch`](https://github.com/Islandora/islandora_newspaper_batch/pull/21/files#diff-bb0eb85e45687085a3b6815a5c75a4a2R496)

... Is there an issue with this?

**Comment by DiegoPino** [comment]
@adam-vessey, I have to read your answer with more dedication to answer back correctly, but it seems you understood my issue. Its all about Using RELS-EXT as a flag/control structure for generating stuff, which in my personal opinion is something we have done always wrong, we could have invented by now a new rdf datastreams just for that purpouse!, but i also agree this pull request is not the place to change it, but also, maybe time to reflect on this and stop doing it.

Side note (funny!) I can't believe that dura space meant that really (probably bad documentation) 

> [RELS-EXT] is intended to be dedicated to solely object-to-object relationships and not used to make general descriptive assertions about objects

when writing fedora 3.x, or the code would simply drop/not consider data properties at all and only leave triples that link info/fedora resources together. RDF is RDF in any place of the world/code if it want to be called RDF and object to object or object to data is RDF, don't think that is the question here.

So back, my question is about A) more versions on RELS-EXT and B) the use of control flags that have no meaning in the objects at all outside our internal derivative processing. C) adding more fake RDF properties to our ontologies. I use the term fake quite liberal here, denoting that their definition makes no sense outside our local environment nor is defined clearly as such. (we even have incremental named rdf properties... never seen that in any other environment).

I feel out derivative process has ways of avoiding more flags. If the setting for generating OCR/HOCR is global, why do we have to keep the state in the object at all?

one extra version in rels ext == hitting a lot of backend. 

Let me check this (test really) because there is a lot of code, good code here and its way better than the huge batch update of the initial commit (thanks @qadan!).
 Did you also consider changing the logic of the origila OCR flag and make it tri-state/multi-state in a single predicate? instead of introducing a new predicate?

File: islandora_ocr.module
Comment (adam-vessey) [reply islandora_ocr.module]: This one yeah... not sure... It [used to be used](https://github.com/Islandora/islandora_ocr/pull/81/commits/d4c4adf45ad4aa101d29dcd8ab9001a078ffeb12), but things were squashed... May be an oversight? Or may just need a comment? @qadan / @willtp87 ?

**Comment by adam-vessey** [comment]
@DiegoPino: The generation is not based solely the global config... As hinted at with [the calls for setting the flags from the batches](https://github.com/Islandora/islandora_ocr/pull/81#discussion_r136589182), it is exposed per batch... Hypothetically, it would be possible to generate objects for every batch and store the necessary configuration there; however, it would be necessary to have all the derivative code be aware of it also... not at all impossible, but an example of CPU-time/storage tradeoffs in dereferencing the configuration vs storing it locally. Additionally, I suspect there may also be individual ingest step forms implemented in various places exposing the object-level flag functionality.

As for overloading/extending the original predicate: I believe this was considered; however, with the possibility of code existing in the wild which may make reference to current values explicitly, I suspect it was elected to go with another predicate to reduce how much things might break.

**Comment by rosiel** [comment]
Is putting object-level configuration settings in RELS-EXT bad? Probably.

But in re. this PR, since the generate_ocr and generate_hocr relationships take TRUE or FALSE as their literal values, why can't we avoid the undesirable repository-wide update by making the new predicate optional (though will be set on all new objects) - i.e. using the value of generate_ocr if generate_hocr is not present?

File: islandora_ocr.module
Comment (DiegoPino) [inline islandora_ocr.module]: Will wait for this legacy piece (this is 1/2 "ISLANDORA_OCR_UPDATE_7001_TIMESTAMP" references) to be removed before re-testing.

**Review by DiegoPino — CHANGES_REQUESTED** [review]
remove legacy code from previous commit

**Comment by qadan** [comment]
Whew! Legacy line removed. Sorry that took me a while :/

In an effort to further assuage the concerns about this pull, https://github.com/Islandora/islandora_paged_content/pull/131 has been updated as well; generating statements have been moved from the derivative code to their own derivative in paged content. A helper has been added which removes them both without auto-committing; putting all this together, we should end up with the same amount of RELS-EXT versions (before and after) as we did before.

@rosiel I've added in a [check during derivative generation](https://github.com/qadan/islandora_ocr/blob/e22bb93bf0c2453605a86fd4d1439cceefd0223b/includes/derivatives.inc#L281-L288) to ensure old behaviour is maintained

**Comment by rosiel** [comment]
@qadan thanks for the check during derivative generation. Here's to no unnecessary RELS-EXT versions.

File: islandora_ocr.install
Comment (rosiel) [inline islandora_ocr.install]: Legacy?

Comment (qadan) [reply islandora_ocr.install]: Whoops, good catch. Removed

**Comment by jordandukart** [comment]
Bump @Islandora/7-x-1-x-committers, @dannylamb

File: includes/derivatives.inc
Comment (DiegoPino) [inline includes/derivatives.inc:255]: Just in case, can you make a type default? How do you feel about that?

File: includes/derivatives.inc
Comment (DiegoPino) [inline includes/derivatives.inc:347]: I see, there is a wrapper so maybe the default for type is not really needed

File: includes/derivatives.inc
Comment (DiegoPino) [inline includes/derivatives.inc:499]: Question. Lets say for N reason the one/both derivatives where no generate(failure, wrong tesseract path) can this function check somehow/get status on that before the removing the flag? Could the prdicate based flag be actually used to re-try?

**Review by DiegoPino — APPROVED** [review]
Approving but left some comments there if you could give them a look? Good work, love how this evolved.

File: includes/derivatives.inc
Comment (qadan) [reply includes/derivatives.inc:347]: Yes, not sure? It was broken out because the original "derive OCR" actually derives both, which doesn't really work if you want to pick which one you're deriving.

File: includes/derivatives.inc
Comment (qadan) [reply includes/derivatives.inc:499]: I _think_ (may be very wrong) that the intent of the API as it currently exists is to flag in all cases - if derivatives don't get generated due to some kind of error, it gets `watchdog()`ed and the flag is removed, perhaps with the intent that you're going to have to go in there again and retry anyway, at which time I don't _think_ the flag is required since the manual derivative generation passes on the `$force` flag? This way you're not setting/unsetting/setting/unsetting the flag each time you want to regenerate and making a ton of versions

**Comment by jordandukart** [comment]
@DiegoPino

**Comment by DiegoPino** [comment]
@jordandukart @qadan putting a note in my calendar. Can go back to testing/reviewing all the changes   on Thursday, is that good for you? It will take me that day like 2 hours but should be done.

**Comment by jordandukart** [comment]
@DiegoPino, @dannylamb, @Islandora/7-x-1-x-committers

**Comment by DiegoPino** [comment]
Got this @jordandukart, will be done by noon

**Comment by DiegoPino** [comment]
@qadan @jordandukart this looks good to me. What is the order of merge/dependencies here? If we put this in, do we need to update all those others together?

**Comment by qadan** [comment]
@DiegoPino whew, thanks for this! Sorry, GitHub is giving me issues with actually notifying me of things.

The order of operations is: https://github.com/Islandora/islandora_ocr/pull/81, then https://github.com/Islandora/islandora_paged_content/pull/131. Then, https://github.com/Islandora/islandora_solution_pack_book/pull/196 https://github.com/Islandora/islandora_book_batch/pull/68 and https://github.com/Islandora/islandora_newspaper_batch/pull/21 all use the new config; those are any-order type things

**Comment by DiegoPino** [comment]
@qadan great. Will get your things merged in that order in 2 hours and 15  minutes more (getting into a meeting now). Just to be sure, we don't have any tests that would need to be updated also??

**Comment by qadan** [comment]
Issues resolved in the associated PR to paged content. I don't believe changes to the tests are required here; there aren't any ingest tests for this, which is what the changes are regarding

**Pull request merged**
